# Examen u2 Lorenzo ismael jimenez lara y Ramon Ochoa Mendez 

In [122]:
#importamos las librerias generales  
import matplotlib.pyplot as plt 
import pandas as pd 
import numpy as np


# parte uno, modelo de regresion 

In [135]:
#librerias para el modelo de regresion lineal 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split


In [124]:
df = pd.read_csv('DataSet1.csv')
df2= df[['Depthm','T_degC']]

df2

C:\Users\Usuario\AppData\Local\Temp\ipykernel_16656\256404182.py:1: DtypeWarning: Columns (47,73) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('DataSet1.csv')


Depthm  T_degC
0            0  10.500
1            8  10.460
2           10  10.460
3           19  10.450
4           20  10.450
...        ...     ...
864858       0  18.744
864859       2  18.744
864860       5  18.692
864861      10  18.161
864862      15  17.533

[864863 rows x 2 columns]

In [140]:
# vemos cual es la correlacion que tienen estas dos columnas 
correlacion = df['Depthm'].corr(df['T_degC'])
correlacion 

-0.6812014451784353

In [169]:
df.isna().sum()

Cst_Cnt                     0
Btl_Cnt                     0
Sta_ID                      0
Depth_ID                    0
Depthm                      0
                        ...  
TA1                    862779
TA2                    864629
pH2                    864853
pH1                    864779
DIC Quality Comment    864808
Length: 74, dtype: int64

In [175]:
# creacion del modelo base 
from sklearn import linear_model

# al entrenar el modelo me aparecia un erro que la entrada conteni NaN perp al verificarlo con isnull parecia que no existina valores nulos,
# aun asi tuve que rellenarlos con la media para poder entrenar el modelo
col1 = df['Depthm'].mean()
df['Depthm'].fillna(col1, inplace=True)
col2 = df['T_degC'].mean()
df['T_degC'].fillna(col2, inplace=True)

#despues de esto ya  me permite entrenar el modelo 

X_train, X_test, y_train, y_test = train_test_split(df['Depthm'], df['T_degC'], test_size=0.2)
model = LinearRegression()
model.fit(X_train.values.reshape(-1, 1), y_train)

LinearRegression()

# breve descripcion del metodo 
la regression lineal es un metodo que se utliza para modelar la relacion entre una variable dependiente y una o mas variables independientes el objetivo pirncipal se puede decir que es encontrar una linea recta que mejor se ajuste a los datos 

In [179]:
# realizar la validacion cruzada 
from sklearn.model_selection import cross_val_score

X = df['Depthm'].values.reshape(-1, 1)
y = df['T_degC']

lr = LinearRegression()
scores = cross_val_score(lr, X, y, cv=5)
print('Precisión de la validación cruzada:', scores.mean())


Precisión de la validación cruzada: 0.4308761468266649


In [187]:
#media y desviacion estandar



# por que es importante relizar la validacion cruzada 

porque nos brinda una estimacion mas precisa del rendimiento de nustro modelo.

# segun internet el modelo de regresion lienal no tinene hiperparametros que se puedan modificar, por lo tanto decido dejar hasta aqui esta parte.

# parte 2, modelo de clasificacion 

In [81]:
# importamos la librerias para el arbol de decision 
from sklearn import preprocessing
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
import pandas as pd 

In [ ]:
# importamos los datos para el arbol de decision, este caso es el data2
Data2 = pd.read_csv('data2.csv')
Data2


In [83]:
X = Data2.iloc[:,:-1]
y = Data2[["Intensity"]]

In [84]:
#entrenamos los datos 
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.6,random_state=45)

In [85]:
# creamos el modelo basico 
modelo_dt = RandomForestClassifier()
modelo_dt.fit(X_train, y_train)
modelo_dt.score(X_test, y_test)

C:\Users\Usuario\AppData\Local\Temp\ipykernel_16656\2369217145.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  modelo_dt.fit(X_train, y_train)


0.9875

# breve descripcion del algoritmo y suposiciones 
se utiliza el arbol de desicion cuando queremos clasificar nuevo datos y el resultado que se mustra es la desicion que se tomo sobre cual nodo es el que se debe elegir

In [87]:
# validacion cruzada infromando la media y desviacion estandar 
from sklearn.model_selection import cross_val_score
from sklearn.metrics import f1_score
from sklearn.model_selection import GridSearchCV

grid = {
    'n_estimators' : [200,300,400,500],
    'max_features' : ['sqrt','log2'],
    'max_depth'    : [4,5,6,7,8],
    'criterion'    : ['gini','entropy'],
    'random_state': [18] 
    
}
rf_cv = GridSearchCV(estimator = RandomForestClassifier(), param_grid = grid, cv=5)
rf_cv.fit(X_train, y_train)


scores = cross_val_score(modelo_dt, X, y, cv=5)

print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))




c:\Users\Usuario\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\model_selection\_validation.py:686: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
c:\Users\Usuario\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\model_selection\_validation.py:686: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
c:\Users\Usuario\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\model_selection\_validation.py:686: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
c:\Users\Usuario\AppData\Local\Pro

Accuracy: 0.98 (+/- 0.01)


# explicacion de porque es importante la validacion cruzada 

la validacion cruzada nos permite evaluar el rendimiento de una forma un tanto mas precisa y confiable que dividir los datos en un conjunto de entrenamiento y uno de prueba, es de alguna manera mas confiable al predecir

In [ ]:
# verificamos cuales son los mejores parmetros para ajustarlos 
rf_cv.best_params_

In [89]:
#ajustamos los nuevos parametros 
rf2 = RandomForestClassifier(n_estimators = 200,
                             max_depth    = 4,
                             max_features = 'sqrt',
                             random_state = 18,
                             criterion    = 'gini').fit(X_train,np.ravel(y_train))

In [ ]:
# puntuacion de la validacion cruzada 
y_pred = rf2.predict(X_test)
accuracy_score(y_test, y_pred)

In [ ]:
confusion_matrix(y_test, y_pred)

# aplique el significado de las metricas de precision, recupreacion y puntaje F1

son meidas comunes que se utilizan ppara evaluar el rendimiento de un modelo de clasificacion, estas metricas se calculan a partir de la matriz de confusión, que es una tabla que resume el numero de verdaderos positivos, verdaderos negativos, falsos positivos y falsos negativos que produce el modelo.



In [ ]:
# modelo ajustado con hiperparametros 
modelo_dt = RandomForestClassifier(criterion= "gini",max_depth=4,max_features="sqrt",n_estimators= 200,random_state = 18)
modelo_dt.fit(X_train, y_train)
modelo_dt.score(X_test, y_test)

# comparacion del modelo base con el ajustado
en el modelo base tenemos un reslutado de 0.9875 y en el nuevo modelo ajustado tenemos un resultado de 0.98125,
